File: leetcode/heap/23. Merge k Sorted Lists/cpp/main.cpp
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

struct Cmp {
    bool operator()(const ListNode *a, const ListNode *b) {
        return a->val > b->val;
    }
};

// Time O(NlogK)
// Space O(K)
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, Cmp> pq;
        
        for(auto list : lists) {
            if (list) {
                pq.push(list);
            }
        }
        
        ListNode dummy(0), *tail = &dummy;
        
        while (!pq.empty()) {
            auto node = pq.top();
            pq.pop();
            
            if (node->next) {
                pq.push(node->next);
            }
            
            tail->next = node;
            tail = tail->next;
        }        
        
        return dummy.next;
    }
};

File: leetcode/heap/23. Merge k Sorted Lists/js/main.js
// Time:
// O(N * Log * K) где k - количество связанных списков
// Стоимость сравнения будет уменьшена до O(Log k) для каждого удаления и вставки в приоритетную очередь.
// Но нахождение узла с наименьшим значением O(1) времени.
// В окончательном связанном списке есть N узлов.

// Space:
// Создание нового связанного списка стоит O(N)
// В приведенном выше коде применяется метод на месте, стоимость которого O (1).
// А очередь с приоритетами (часто реализуемая с помощью куч) стоит O (k) пространства
// В большинстве случаев она намного меньше, чем N).
const mergeKLists = lists => {
  let pq = new PriorityQueue({ comparator: (a, b) => a.val - b.val });

  for (let list of lists) {
    if (list) {
      pq.offer(list);
    }
  }

  let dummy = new ListNode();
  let p = dummy;

  while (pq.size() > 0) {
    let list = pq.poll();

    if (list && list.next) {
      pq.offer(list.next);
    }

    p.next = list;
    p = p.next;
  }

  return dummy.next;
};
